From 5c309d9ff97a3027008a628dae803b1746ef5b4c Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Wed, 21 May 2025 18:09:33 +0300
Subject: [PATCH] [JSW-703] jgs: export arc region data to specs

with this patch, removed arc region declerations
from arcboot to specs for more visibility
to next stage firmware. here after any memory
region base address declerations can be defined
from arc region declarations from soc header.

Change-Id: I58973ebeb388bddf99c90591dae8d8ac8417bf3c
---

diff --git a/arch/fs1/arc/dev_manifest.h b/arch/fs1/arc/dev_manifest.h
new file mode 100644
index 0000000..f014dc3
--- /dev/null
+++ b/arch/fs1/arc/dev_manifest.h
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2025 Intel, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef _DEV_BOOT_DESC_H_
+#define _DEV_BOOT_DESC_H_
+
+#include <fs1_blocks_linux_driver.h>
+#include <address_space.h>
+
+#define SPI_MANIFEST_IMG_0_FLASH_OFFSET		(0x0)
+/* manifest 0 boot pointers are DMA trans to MNG ARC ICCM at offset 0 */
+#define SPI_MANIFEST_IMG_0_ICCM_OFFSET		(0)
+#define SPI_MANIFEST_IMG_0_DMA_DEST			((IO_0_MNG_ARC_ICCM_BASE_FULL) + (SPI_MANIFEST_IMG_0_ICCM_OFFSET))
+/* 32Bit arc viewpoint address of manifest 0 stored in ICCM */
+#define SPI_MANIFEST_IMG_0_DMA_DEST_ARC		((IO_0_MNG_ARC_ICCM_BASE) + (SPI_MANIFEST_IMG_0_ICCM_OFFSET))
+
+
+#define SPI_MANIFEST_IMG_1_FLASH_OFFSET		(0x200000)
+/* manifest 1 boot pointers are DMA trans to MNG ARC ICCM at offset 64K */
+#define SPI_MANIFEST_IMG_1_ICCM_OFFSET		SZ_64K
+#define SPI_MANIFEST_IMG_1_DMA_DEST			((IO_0_MNG_ARC_ICCM_BASE_FULL) + (SPI_MANIFEST_IMG_1_ICCM_OFFSET))
+/* 32Bit arc viewpoint address of manifest 1 stored in ICCM */
+#define SPI_MANIFEST_IMG_1_DMA_DEST_ARC		((IO_0_MNG_ARC_ICCM_BASE) + (SPI_MANIFEST_IMG_1_ICCM_OFFSET))
+
+
+/* For Fake fuse manifest */
+#define SPI_MANIFEST_IMG_2_ICCM_OFFSET		SZ_128K
+#define SPI_MANIFEST_IMG_2_DMA_DEST			((IO_0_MNG_ARC_ICCM_BASE_FULL) + (SPI_MANIFEST_IMG_2_ICCM_OFFSET))
+/* 32Bit arc viewpoint address of manifest 2 stored in ICCM */
+#define SPI_MANIFEST_IMG_2_DMA_DEST_ARC		((IO_0_MNG_ARC_ICCM_BASE) + (SPI_MANIFEST_IMG_2_ICCM_OFFSET))
+
+#endif /* _DEV_BOOT_DESC_H_ */
diff --git a/arch/fs1/arc/dev_security.h b/arch/fs1/arc/dev_security.h
index 733d584..e633717 100644
--- a/arch/fs1/arc/dev_security.h
+++ b/arch/fs1/arc/dev_security.h
@@ -60,8 +60,8 @@
 							(mmPSOC_SECURITY_MIN_FW_VERSION - 1)
 
 /* RECOVERY GEN_MEM */
-#define BOOT_ROM_REC_RR_GEN_MEM_MIN		(HL_FLASH_BASE)
-#define BOOT_ROM_REC_RR_GEN_MEM_MAX		(HL_FLASH_BASE) + (HL_FLASH_SIZE - 1)
+#define BOOT_ROM_REC_RR_GEN_MEM_MIN		(SPI_FLASH_BASE)
+#define BOOT_ROM_REC_RR_GEN_MEM_MAX		(SPI_FLASH_BASE) + (SPI_FLASH_SIZE - 1)
 
 /* SAI Table start */
 #define SECURITY_PCI_LBW_SAI_VALUE		2
diff --git a/arch/fs1/arc/dev_spi_flash.c b/arch/fs1/arc/dev_spi_flash.c
index efb317a..f38c931 100644
--- a/arch/fs1/arc/dev_spi_flash.c
+++ b/arch/fs1/arc/dev_spi_flash.c
@@ -45,7 +45,7 @@
 {
 	/* mask offset to valid range of flash */
 	offset &= FLASH_VALID_MASK;
-	uint32_t val = base_readl(HL_FLASH_BASE + offset);
+	uint32_t val = base_readl(SPI_FLASH_BASE + offset);
 
 	return val;
 }
diff --git a/arch/fs1/arc/soc.h b/arch/fs1/arc/soc.h
index 570c622..cfd0c6e 100644
--- a/arch/fs1/arc/soc.h
+++ b/arch/fs1/arc/soc.h
@@ -6,6 +6,9 @@
 #ifndef _SOC_H_
 #define _SOC_H_
 
+#define MASK_32_BIT			0xFFFFFFFF
+#define MASK_28_BIT			0xFFFFFFF
+
 enum region_idx {
 	REGION_ICCM_IDX = 0,
 	REGION_HBW_0_IDX = 1,
@@ -25,28 +28,20 @@
 	REGION_LBW_CFG_IDX = 15,
 };
 
-enum gpio_func {
-	GPIO_FUNC_0,
-	GPIO_FUNC_1,
-	GPIO_FUNC_2,
-	GPIO_FUNC_3,
-};
-
 #define REGION_BASE_ADDR(REG_IDX)	(REG_IDX << 28)
 #define SZ_16				0x00000010
 #define SZ_32K				0x00008000
 #define SZ_64K				0x00010000
 #define SZ_128K				0x00020000
 #define SZ_256K				0x00040000
-#define SZ_1MB				0x00100000
 #define SZ_4MB				0x00400000
 #define SZ_256MB			0x10000000
 #define SZ_1G				0x40000000
 
 #define FLASH_VALID_MASK		0x1FFFFF
 
-#define ARC_REG_BASE_MASK	REGION_BASE_ADDR(0xF)
-#define HL_CFG_BASE			REGION_BASE_ADDR(REGION_LBW_CFG_IDX)
+#define ARC_REGION_MASK			REGION_BASE_ADDR(0xF)
+#define CFG_SPACE_REG_BASE		REGION_BASE_ADDR(REGION_LBW_CFG_IDX)
 #define HL_CFG_SIZE			SZ_256MB   /* 256MB */
 
 #define HL_ICCM_BASE			REGION_BASE_ADDR(REGION_ICCM_IDX)
@@ -60,9 +55,11 @@
 #define HL_FW_MEM_BASE			(HL_FW_MEM_REGION_BASE + HL_FW_MEM_OFFSET)
 #define HL_FW_MEM_SIZE			SZ_4MB
 
-#define HL_GEN_MEM_BASE			REGION_BASE_ADDR(REGION_GEN_MEM_IDX)
-#define HL_FLASH_OFFSET			0xF000000
-#define HL_FLASH_SIZE			(16 * SIZE_MB)
+#define GEN_MEM_REG_BASE		REGION_BASE_ADDR(REGION_GEN_MEM_IDX)
+
+#define SPI_FLASH_OFFSET		0xF000000
+#define SPI_FLASH_SIZE			(16 * SZ_1MB)
+#define SPI_FLASH_BASE			(GEN_MEM_REG_BASE + SPI_FLASH_OFFSET)
 
 /* Below definitions are applicable to IO Die 0 only */
 #define RST_ARC_ICCM_OFFSET		0x0000000
@@ -94,6 +91,8 @@
 #define PCIE_FWSRAM_OFFSET		0x0380000
 #define PCIE_FWSRAM_SIZE		SZ_64K
 
+#define GET_FULL_ADDR(REG_OFF)		((REG_OFF & MASK_28_BIT) + CFG_MEM_FULL_BASE)
+
 #define GEN_MEM_FULL_BASE		0x300007f00000000ull
 #define CFG_MEM_FULL_BASE		0x300007ff0000000ull
 #define FULL_ADDR_MSB			0x300007F
@@ -119,27 +118,8 @@
 #define CONV_ARC_REG_TO_GEN_MEM(addr)	(((addr) & 0xFFFFFFF) + ((GEN_MEM_FULL_BASE) & 0xFFFFFFFF))
 #define CONV_ARC_REG_TO_CFG_MEM(addr)	(((addr) & 0xFFFFFFF) + ((CFG_MEM_FULL_BASE) & 0xFFFFFFFF))
 
-#define SPI_MANIFEST_IMG_0_FLASH_OFFSET		(0x0)
-/* manifest 0 boot pointers are DMA trans to MNG ARC ICCM at offset 0 */
-#define SPI_MANIFEST_IMG_0_ICCM_OFFSET		(0)
-#define SPI_MANIFEST_IMG_0_OFFSET		(MNG_ARC_ICCM_OFFSET + SPI_MANIFEST_IMG_0_ICCM_OFFSET)
-#define SPI_MANIFEST_IMG_0_DMA_DEST		(FW_MEM_BASE_FULL + SPI_MANIFEST_IMG_0_OFFSET)
-/* 32Bit arc viewpoint address of manifest 0 stored in ICCM */
-#define SPI_MANIFEST_IMG_0_DMA_DEST_ARC		(HL_FW_MEM_BASE + SPI_MANIFEST_IMG_0_OFFSET)
 
-#define SPI_MANIFEST_IMG_1_FLASH_OFFSET		(0x200000)
-/* manifest 1 boot pointers are DMA trans to MNG ARC ICCM at offset 64K */
-#define SPI_MANIFEST_IMG_1_ICCM_OFFSET		SZ_64K
-#define SPI_MANIFEST_IMG_1_OFFSET		(MNG_ARC_ICCM_OFFSET + SPI_MANIFEST_IMG_1_ICCM_OFFSET)
-#define SPI_MANIFEST_IMG_1_DMA_DEST		(FW_MEM_BASE_FULL + SPI_MANIFEST_IMG_1_OFFSET)
-/* 32Bit arc viewpoint address of manifest 1 stored in ICCM */
-#define SPI_MANIFEST_IMG_1_DMA_DEST_ARC		(HL_FW_MEM_BASE + SPI_MANIFEST_IMG_1_OFFSET)
-
-/* For Fake fuse manifest */
-#define SPI_MANIFEST_IMG_2_ICCM_OFFSET	SZ_128K
-#define SPI_MANIFEST_IMG_2_OFFSET		(MNG_ARC_ICCM_OFFSET + SPI_MANIFEST_IMG_2_ICCM_OFFSET)
-#define SPI_MANIFEST_IMG_2_DMA_DEST		(FW_MEM_BASE_FULL + SPI_MANIFEST_IMG_2_OFFSET)
-/* 32Bit arc viewpoint address of manifest 2 stored in ICCM */
-#define SPI_MANIFEST_IMG_2_DMA_DEST_ARC		(HL_FW_MEM_BASE + SPI_MANIFEST_IMG_2_OFFSET)
+#define IO_0_MNG_ARC_ICCM_BASE_FULL	(FW_MEM_BASE_FULL + MNG_ARC_ICCM_OFFSET)
+#define IO_0_MNG_ARC_ICCM_BASE		(CONV_64B_TO_ARC_FW_MEM(IO_0_MNG_ARC_ICCM_BASE_FULL))
 
 #endif /* _SOC_H_ */
diff --git a/arch/jgs/arc/common.c b/arch/jgs/arc/common.c
index 2643e80..1ef91ac 100644
--- a/arch/jgs/arc/common.c
+++ b/arch/jgs/arc/common.c
@@ -95,7 +95,7 @@
 #define HL_ARC1_DCCM_PERSIST_SECTION_SIZE		(3 * 1024) /* 3KB */
 void boot_rom_persistent_dccm_init(uint8_t die_id)
 {
-	volatile uint32_t *dccm_addr = (uint32_t *) HL_DCCM_BASE;
+	volatile uint32_t *dccm_addr = (uint32_t *) LOCAL_DCCM_BASE;
 	uint32_t dccm_dwords = HL_ARC1_DCCM_PERSIST_SECTION_SIZE >> 2;
 	struct fs1_cold_rst_data cold_rst_data;
 	uint32_t i;
@@ -194,7 +194,7 @@
 	if (cold_rst_data.arc1_iccm_init_done)
 		return;
 
-	ret = boot_rom_mem_init(MNG_ARC_ICCM_FULL_BASE, MNG_ARC_ICCM_SIZE);
+	ret = boot_rom_mem_init(IO_0_MNG_ARC_ICCM_BASE_FULL, IO_0_MNG_ARC_ICCM_SIZE);
 	if (ret) {
 		boot_rom_status_update(PROGRESS_INIT_ARC1_ICCM_INIT_FAILED);
 		return;
@@ -364,32 +364,32 @@
 /* This function is only applicable for io die master */
 bool boot_rom_is_valid_address_range(uint64_t addr, uint32_t len)
 {
-	if (addr >= (uint64_t)SCRATCHPAD_SRAM_BASE &&
-	    addr + len <= (uint64_t)(SCRATCHPAD_SRAM_BASE + SCRATCHPAD_SRAM_SIZE))
+	if (addr >= (uint64_t)IO_0_SP_SRAM_BASE &&
+	    addr + len <= (uint64_t)(IO_0_SP_SRAM_BASE + IO_0_SP_SRAM_SIZE))
 		return true;
 
-	if (addr >= SCRATCHPAD_SRAM_BASE_FULL &&
-	    addr + len <= SCRATCHPAD_SRAM_BASE_FULL + SCRATCHPAD_SRAM_SIZE)
+	if (addr >= IO_0_SP_SRAM_BASE_FULL &&
+	    addr + len <= IO_0_SP_SRAM_BASE_FULL + IO_0_SP_SRAM_SIZE)
 		return true;
 
-	if (addr >= RST_ARC_ICCM_FULL_BASE &&
-	    addr + len <= RST_ARC_ICCM_FULL_BASE + RST_ARC_MEM_SIZE)
+	if (addr >= IO_0_RST_ARC_ICCM_BASE_FULL &&
+	    addr + len <= IO_0_RST_ARC_ICCM_BASE_FULL + IO_0_RST_ARC_MEM_SIZE)
 		return true;
 
-	if (addr >= MNG_ARC_ICCM_FULL_BASE &&
-	    addr + len <= MNG_ARC_ICCM_FULL_BASE + MNG_ARC_MEM_SIZE)
+	if (addr >= IO_0_MNG_ARC_ICCM_BASE_FULL &&
+	    addr + len <= IO_0_MNG_ARC_ICCM_BASE_FULL + IO_0_MNG_ARC_MEM_SIZE)
 		return true;
 
-	if (addr >= PID_ARC_ICCM_FULL_BASE &&
-	    addr + len <= PID_ARC_ICCM_FULL_BASE + PID_ARC_MEM_SIZE)
+	if (addr >= IO_0_PID_ARC_ICCM_BASE_FULL &&
+	    addr + len <= IO_0_PID_ARC_ICCM_BASE_FULL + IO_0_PID_ARC_MEM_SIZE)
 		return true;
 
-	if (addr >= (uint64_t)(HL_FW_MEM_BASE + PSOC_FWSRAM_OFFSET) &&
-	    addr + len <= (uint64_t)(HL_FW_MEM_BASE + PSOC_FWSRAM_OFFSET + PSOC_FWSRAM_SIZE))
+	if (addr >= (uint64_t)(IO_0_PSOC_FW_SRAM_BASE) &&
+	    addr + len <= (uint64_t)(IO_0_PSOC_FW_SRAM_BASE + IO_0_FWSRAM_SIZE))
 		return true;
 
-	if (addr >= PSOC_FW_SRAM_FULL_BASE &&
-	    addr + len <= PSOC_FW_SRAM_FULL_BASE + PSOC_FWSRAM_SIZE)
+	if (addr >= IO_0_PSOC_FW_SRAM_BASE_FULL &&
+	    addr + len <= IO_0_PSOC_FW_SRAM_BASE_FULL + IO_0_FWSRAM_SIZE)
 		return true;
 
 	return false;
@@ -397,15 +397,15 @@
 
 uint32_t boot_rom_fw_mem_arc_addr(uint64_t addr)
 {
-	if (addr >= MNG_ARC_ICCM_FULL_BASE && addr <= MNG_ARC_ICCM_FULL_BASE + MNG_ARC_MEM_SIZE)
-		return CONV_64B_TO_ARC_FW_MEM(addr) - (HL_FW_MEM_BASE + MNG_ARC_ICCM_OFFSET);
+	if (addr >= IO_0_MNG_ARC_ICCM_BASE_FULL && addr <= IO_0_MNG_ARC_ICCM_BASE_FULL + IO_0_MNG_ARC_MEM_SIZE)
+		return CONV_64B_TO_ARC_FW_MEM(addr) - (IO_0_MNG_ARC_ICCM_BASE);
 	else
 		return CONV_64B_TO_ARC_FW_MEM(addr);
 }
 
 uint32_t boot_rom_get_arc_addr(uint64_t addr)
 {
-	if (addr >= FW_MEM_BASE_FULL && addr <= FW_MEM_BASE_FULL + HL_FW_MEM_SIZE)
+	if (addr >= (IO_0_FW_MEM_BASE_FULL) && addr <= (IO_0_FW_MEM_BASE_FULL) + IO_0_CFG_FW_MEM_SECTION)
 		return boot_rom_fw_mem_arc_addr(addr);
 
 	return 0;
@@ -450,7 +450,6 @@
  * Request CPLD to start the watchdog, by asserting
  * a GPIO_66 (Boot_WD).
  */
-#define GPIOS_PER_BANK			16
 #define FS1_CPLD_WDOG_GPIO		66
 void boot_rom_start_cpld_wdtimer(void)
 {
diff --git a/arch/jgs/arc/dev_boot_info.h b/arch/jgs/arc/dev_boot_info.h
index 7a547b8..7290f07 100644
--- a/arch/jgs/arc/dev_boot_info.h
+++ b/arch/jgs/arc/dev_boot_info.h
@@ -6,13 +6,13 @@
 #ifndef _DEV_BOOT_INFO_H_
 #define _DEV_BOOT_INFO_H_
 
-#include <soc.h>
-#include <boot_rom.h>
+#include <jgs_blocks_linux_driver.h>
+#include <address_space.h>
 
 #define LFS_TAG_BOOT_INFO		"boot_info_"
 
 #define OFFSET_32KB		(32 * 1024)
-#define CONFIG_BOOTINFO_ADDR	(HL_FW_MEM_BASE + SCRATCHPAD_SRAM_OFFSET)
+#define CONFIG_BOOTINFO_ADDR	(IO_0_SP_SRAM_BASE)
 #define CONFIG_BOOTINFO_SIZE	OFFSET_32KB
 
 void boot_rom_set_boot_info(void);
diff --git a/arch/jgs/arc/dev_dma.c b/arch/jgs/arc/dev_dma.c
index 6ca647d..a34b1a0 100644
--- a/arch/jgs/arc/dev_dma.c
+++ b/arch/jgs/arc/dev_dma.c
@@ -12,6 +12,7 @@
 #include <parc_global_conf_regs.h>
 #include <jgs_reg_map.h>
 #include <dma.h>
+#include <address_space.h>
 
 #define DMA_ADDRESS_LOW(x)	((x) & MASK_32_BIT)
 #define DMA_ADDRESS_HIGH(x)	(((x) >> 32) & MASK_32_BIT)
@@ -21,8 +22,8 @@
 enum dma_dest_type boot_rom_get_dma_dest_type(uint64_t addr)
 {
 	/* TODO: Add checks for all offsets in FW MEM  for other dies */
-	if (addr >= PSOC_FW_SRAM_FULL_BASE &&
-	    addr < (PSOC_FW_SRAM_FULL_BASE + PSOC_FWSRAM_SIZE))
+	if (addr >= IO_0_PSOC_FW_SRAM_BASE_FULL &&
+	    addr < (IO_0_PSOC_FW_SRAM_BASE_FULL + IO_0_FWSRAM_SIZE))
 		return DMA_DEST_SRAM;
 
 	return DMA_DEST_HBW;
diff --git a/arch/jgs/arc/dev_efuse.h b/arch/jgs/arc/dev_efuse.h
index 86153a1..52cbfc0 100644
--- a/arch/jgs/arc/dev_efuse.h
+++ b/arch/jgs/arc/dev_efuse.h
@@ -8,8 +8,6 @@
 
 #include <jgs_efuse.h>
 
-#define FLASH_VALID_MASK			0x1FFFFF
-
 #define FUSE_READ_FSM_TRIGGER		0x1
 #define FUSE_DELAY_US				5
 #define FUSE_TIMEOUT_US				500
diff --git a/arch/jgs/arc/dev_manifest.h b/arch/jgs/arc/dev_manifest.h
new file mode 100644
index 0000000..871614a
--- /dev/null
+++ b/arch/jgs/arc/dev_manifest.h
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2025 Intel, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef _DEV_BOOT_DESC_H_
+#define _DEV_BOOT_DESC_H_
+
+#include <jgs_blocks_linux_driver.h>
+#include <address_space.h>
+
+#define SPI_MANIFEST_IMG_0_FLASH_OFFSET		(0x0)
+/* manifest 0 boot pointers are DMA trans to MNG ARC ICCM at offset 0 */
+#define SPI_MANIFEST_IMG_0_ICCM_OFFSET		(0)
+#define SPI_MANIFEST_IMG_0_DMA_DEST			(GET_FULL_ADDR(mmIO_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE) + (SPI_MANIFEST_IMG_0_ICCM_OFFSET))
+/* 32Bit arc viewpoint address of manifest 0 stored in ICCM */
+#define SPI_MANIFEST_IMG_0_DMA_DEST_ARC		((mmIO_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE) + (SPI_MANIFEST_IMG_0_ICCM_OFFSET))
+
+
+#define SPI_MANIFEST_IMG_1_FLASH_OFFSET		(0x200000)
+/* manifest 1 boot pointers are DMA trans to MNG ARC ICCM at offset 64K */
+#define SPI_MANIFEST_IMG_1_ICCM_OFFSET		SZ_64K
+#define SPI_MANIFEST_IMG_1_DMA_DEST			(GET_FULL_ADDR(mmIO_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE) + (SPI_MANIFEST_IMG_1_ICCM_OFFSET))
+/* 32Bit arc viewpoint address of manifest 1 stored in ICCM */
+#define SPI_MANIFEST_IMG_1_DMA_DEST_ARC		((mmIO_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE) + (SPI_MANIFEST_IMG_1_ICCM_OFFSET))
+
+
+/* For Fake fuse manifest */
+#define SPI_MANIFEST_IMG_2_ICCM_OFFSET		SZ_128K
+#define SPI_MANIFEST_IMG_2_DMA_DEST			(GET_FULL_ADDR(mmIO_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE) + (SPI_MANIFEST_IMG_2_ICCM_OFFSET))
+/* 32Bit arc viewpoint address of manifest 2 stored in ICCM */
+#define SPI_MANIFEST_IMG_2_DMA_DEST_ARC		((mmIO_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE) + (SPI_MANIFEST_IMG_2_ICCM_OFFSET))
+
+#endif /* _DEV_BOOT_DESC_H_ */
diff --git a/arch/jgs/arc/dev_security.h b/arch/jgs/arc/dev_security.h
index 01a3cdf..9cec429 100644
--- a/arch/jgs/arc/dev_security.h
+++ b/arch/jgs/arc/dev_security.h
@@ -60,8 +60,8 @@
 							(mmPSOC_SECURITY_MIN_FW_VERSION - 1)
 
 /* RECOVERY GEN_MEM */
-#define BOOT_ROM_REC_RR_GEN_MEM_MIN		(HL_FLASH_BASE)
-#define BOOT_ROM_REC_RR_GEN_MEM_MAX		(HL_FLASH_BASE) + (HL_FLASH_SIZE - 1)
+#define BOOT_ROM_REC_RR_GEN_MEM_MIN		(SPI_FLASH_BASE)
+#define BOOT_ROM_REC_RR_GEN_MEM_MAX		(SPI_FLASH_BASE) + (SPI_FLASH_SIZE - 1)
 
 /* SAI Table start */
 #define SECURITY_PCI_LBW_SAI_VALUE		2
diff --git a/arch/jgs/arc/dev_spi_flash.c b/arch/jgs/arc/dev_spi_flash.c
index 3c0613a..525729a 100644
--- a/arch/jgs/arc/dev_spi_flash.c
+++ b/arch/jgs/arc/dev_spi_flash.c
@@ -45,7 +45,7 @@
 {
 	/* mask offset to valid range of flash */
 	offset &= FLASH_VALID_MASK;
-	uint32_t val = base_readl(HL_FLASH_BASE + offset);
+	uint32_t val = base_readl(SPI_FLASH_BASE + offset);
 
 	return val;
 }
diff --git a/arch/jgs/arc/soc.h b/arch/jgs/arc/soc.h
deleted file mode 100644
index 570c622..0000000
--- a/arch/jgs/arc/soc.h
+++ /dev/null
@@ -1,145 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2024 HabanaLabs Ltd.
- */
-
-#ifndef _SOC_H_
-#define _SOC_H_
-
-enum region_idx {
-	REGION_ICCM_IDX = 0,
-	REGION_HBW_0_IDX = 1,
-	REGION_HBW_1_IDX = 2,
-	REGION_HBW_2_IDX = 3,
-	REGION_HBW_3_IDX = 4,
-	REGION_HBW_4_IDX = 5,
-	REGION_HBW_5_IDX = 6,
-	REGION_FW_MEM_IDX = 7,
-	REGION_DCCM_IDX = 8,
-	REGION_GEN_MEM_IDX = 9,
-	REGION_GT_TILE_0_IDX = 10,
-	REGION_GT_TILE_1_IDX = 11,
-	REGION_GT_TILE_2_IDX = 12,
-	REGION_GT_TILE_3_IDX = 13,
-	REGION_LBW_DBG_IDX = 14,
-	REGION_LBW_CFG_IDX = 15,
-};
-
-enum gpio_func {
-	GPIO_FUNC_0,
-	GPIO_FUNC_1,
-	GPIO_FUNC_2,
-	GPIO_FUNC_3,
-};
-
-#define REGION_BASE_ADDR(REG_IDX)	(REG_IDX << 28)
-#define SZ_16				0x00000010
-#define SZ_32K				0x00008000
-#define SZ_64K				0x00010000
-#define SZ_128K				0x00020000
-#define SZ_256K				0x00040000
-#define SZ_1MB				0x00100000
-#define SZ_4MB				0x00400000
-#define SZ_256MB			0x10000000
-#define SZ_1G				0x40000000
-
-#define FLASH_VALID_MASK		0x1FFFFF
-
-#define ARC_REG_BASE_MASK	REGION_BASE_ADDR(0xF)
-#define HL_CFG_BASE			REGION_BASE_ADDR(REGION_LBW_CFG_IDX)
-#define HL_CFG_SIZE			SZ_256MB   /* 256MB */
-
-#define HL_ICCM_BASE			REGION_BASE_ADDR(REGION_ICCM_IDX)
-#define HL_ICCM_SIZE			0x40000   /* 256KB */
-
-#define HL_DCCM_BASE			REGION_BASE_ADDR(REGION_DCCM_IDX)
-#define HL_DCCM_SIZE			0x20000   /* 128KB */
-
-#define HL_FW_MEM_REGION_BASE		REGION_BASE_ADDR(REGION_FW_MEM_IDX)
-#define HL_FW_MEM_OFFSET		0x8800000
-#define HL_FW_MEM_BASE			(HL_FW_MEM_REGION_BASE + HL_FW_MEM_OFFSET)
-#define HL_FW_MEM_SIZE			SZ_4MB
-
-#define HL_GEN_MEM_BASE			REGION_BASE_ADDR(REGION_GEN_MEM_IDX)
-#define HL_FLASH_OFFSET			0xF000000
-#define HL_FLASH_SIZE			(16 * SIZE_MB)
-
-/* Below definitions are applicable to IO Die 0 only */
-#define RST_ARC_ICCM_OFFSET		0x0000000
-#define RST_ARC_ICCM_SIZE		SZ_64K
-#define RST_ARC_DCCM_OFFSET		0x0010000
-#define RST_ARC_DCCM_SIZE		SZ_32K
-#define MNG_ARC_ICCM_OFFSET		0x0040000
-#define MNG_ARC_ICCM_SIZE		SZ_256K
-#define MNG_ARC_DCCM_OFFSET		0x0080000
-#define MNG_ARC_DCCM_SIZE		SZ_128K
-#define PID_ARC_ICCM_OFFSET		0x00C0000
-#define PID_ARC_ICCM_SIZE		SZ_128K
-#define PID_ARC_DCCM_OFFSET		0x00E0000
-#define PID_ARC_DCCM_SIZE		SZ_64K
-#define MISC_ARC_ICCM_OFFSET		0x0100000
-#define MISC_ARC_ICCM_SIZE		SZ_256K
-#define MISC_ARC_DCCM_OFFSET		0x0140000
-#define MISC_ARC_DCCM_SIZE		SZ_64K
-#define SBROM_OFFSET			0x0180000
-#define SBROM_SIZE			SZ_128K
-#define PSOC_FWSRAM_OFFSET		0x01A0000
-#define PSOC_FWSRAM_SIZE		SZ_128K
-#define SCRATCHPAD_SRAM_OFFSET		0x0200000
-#define SCRATCHPAD_SRAM_SIZE		SZ_1MB
-#define CXL_ARC_ICCM_OFFSET		0x0300000
-#define CXL_ARC_ICCM_SIZE		SZ_256K
-#define CXL_ARC_DCCM_OFFSET		0x0340000
-#define CXL_ARC_DCCM_SIZE		SZ_256K
-#define PCIE_FWSRAM_OFFSET		0x0380000
-#define PCIE_FWSRAM_SIZE		SZ_64K
-
-#define GEN_MEM_FULL_BASE		0x300007f00000000ull
-#define CFG_MEM_FULL_BASE		0x300007ff0000000ull
-#define FULL_ADDR_MSB			0x300007F
-#define FW_MEM_BASE_FULL		0x300007FF8800000ull
-#define SCRATCHPAD_SRAM_BASE		(HL_FW_MEM_BASE + SCRATCHPAD_SRAM_OFFSET)
-#define SCRATCHPAD_SRAM_BASE_FULL	(FW_MEM_BASE_FULL + SCRATCHPAD_SRAM_OFFSET)
-#define RST_ARC_ICCM_FULL_BASE		(FW_MEM_BASE_FULL + RST_ARC_ICCM_OFFSET)
-#define MNG_ARC_ICCM_FULL_BASE		(FW_MEM_BASE_FULL + MNG_ARC_ICCM_OFFSET)
-#define PID_ARC_ICCM_FULL_BASE		(FW_MEM_BASE_FULL + PID_ARC_ICCM_OFFSET)
-#define MISC_ARC_ICCM_FULL_BASE		(FW_MEM_BASE_FULL + MISC_ARC_ICCM_OFFSET)
-#define CXL_ARC_ICCM_FULL_BASE		(FW_MEM_BASE_FULL + CXL_ARC_ICCM_OFFSET)
-#define PSOC_FW_SRAM_FULL_BASE		(FW_MEM_BASE_FULL + PSOC_FWSRAM_OFFSET)
-
-#define RST_ARC_MEM_SIZE		(RST_ARC_ICCM_SIZE + RST_ARC_DCCM_SIZE)
-#define MNG_ARC_MEM_SIZE		(MNG_ARC_ICCM_SIZE + MNG_ARC_DCCM_SIZE)
-#define PID_ARC_MEM_SIZE		(PID_ARC_ICCM_SIZE + PID_ARC_DCCM_SIZE)
-#define MISC_ARC_MEM_SIZE		(MISC_ARC_ICCM_SIZE + MISC_ARC_DCCM_SIZE)
-#define CXL_ARC_MEM_SIZE		(CXL_ARC_ICCM_SIZE + CXL_ARC_DCCM_SIZE)
-
-#define CONV_64B_TO_ARC_FW_MEM(addr)	((addr & 0x3FFFFF) + HL_FW_MEM_BASE)
-
-/* This 2 macro's to convert ARC based 32B address to LBW MEM map address of 32B */
-#define CONV_ARC_REG_TO_GEN_MEM(addr)	(((addr) & 0xFFFFFFF) + ((GEN_MEM_FULL_BASE) & 0xFFFFFFFF))
-#define CONV_ARC_REG_TO_CFG_MEM(addr)	(((addr) & 0xFFFFFFF) + ((CFG_MEM_FULL_BASE) & 0xFFFFFFFF))
-
-#define SPI_MANIFEST_IMG_0_FLASH_OFFSET		(0x0)
-/* manifest 0 boot pointers are DMA trans to MNG ARC ICCM at offset 0 */
-#define SPI_MANIFEST_IMG_0_ICCM_OFFSET		(0)
-#define SPI_MANIFEST_IMG_0_OFFSET		(MNG_ARC_ICCM_OFFSET + SPI_MANIFEST_IMG_0_ICCM_OFFSET)
-#define SPI_MANIFEST_IMG_0_DMA_DEST		(FW_MEM_BASE_FULL + SPI_MANIFEST_IMG_0_OFFSET)
-/* 32Bit arc viewpoint address of manifest 0 stored in ICCM */
-#define SPI_MANIFEST_IMG_0_DMA_DEST_ARC		(HL_FW_MEM_BASE + SPI_MANIFEST_IMG_0_OFFSET)
-
-#define SPI_MANIFEST_IMG_1_FLASH_OFFSET		(0x200000)
-/* manifest 1 boot pointers are DMA trans to MNG ARC ICCM at offset 64K */
-#define SPI_MANIFEST_IMG_1_ICCM_OFFSET		SZ_64K
-#define SPI_MANIFEST_IMG_1_OFFSET		(MNG_ARC_ICCM_OFFSET + SPI_MANIFEST_IMG_1_ICCM_OFFSET)
-#define SPI_MANIFEST_IMG_1_DMA_DEST		(FW_MEM_BASE_FULL + SPI_MANIFEST_IMG_1_OFFSET)
-/* 32Bit arc viewpoint address of manifest 1 stored in ICCM */
-#define SPI_MANIFEST_IMG_1_DMA_DEST_ARC		(HL_FW_MEM_BASE + SPI_MANIFEST_IMG_1_OFFSET)
-
-/* For Fake fuse manifest */
-#define SPI_MANIFEST_IMG_2_ICCM_OFFSET	SZ_128K
-#define SPI_MANIFEST_IMG_2_OFFSET		(MNG_ARC_ICCM_OFFSET + SPI_MANIFEST_IMG_2_ICCM_OFFSET)
-#define SPI_MANIFEST_IMG_2_DMA_DEST		(FW_MEM_BASE_FULL + SPI_MANIFEST_IMG_2_OFFSET)
-/* 32Bit arc viewpoint address of manifest 2 stored in ICCM */
-#define SPI_MANIFEST_IMG_2_DMA_DEST_ARC		(HL_FW_MEM_BASE + SPI_MANIFEST_IMG_2_OFFSET)
-
-#endif /* _SOC_H_ */
diff --git a/common/boot_desc.c b/common/boot_desc.c
index 1e9ce57..934209a 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -4,13 +4,11 @@
  * All Rights Reserved.
  */
 
-#include <stdint.h>
 #include <string.h>
-#include <stdbool.h>
-#include <hl_fw_upd_if.h>
 #include <boot_rom.h>
+#include <hl_fw_upd_if.h>
 #include <hl_boot_desc.h>
-#include "boot_desc.h"
+#include <boot_desc.h>
 #include <status.h>
 #include <spi_flash.h>
 #include <dev_spi_flash.h>
@@ -19,6 +17,7 @@
 #include <boot_flow.h>
 #include <dev_boot_info.h>
 #include <lfs_wrapper.h>
+#include <dev_manifest.h>
 
 static struct hl_fw_manifest_ctx *p_manifest_ctx;
 static struct hl_fw_boot_ptrs *p_boot_ptr;
diff --git a/common/dma.c b/common/dma.c
index 37e152e..d7c00a0 100644
--- a/common/dma.c
+++ b/common/dma.c
@@ -65,7 +65,7 @@
 	if (image_size < min_dma_size)
 		return DMA_INVALID_IMG_SIZE;
 
-	if ((flash_offset + image_size) > HL_FLASH_SIZE)
+	if ((flash_offset + image_size) > SPI_FLASH_SIZE)
 		return DMA_INVALID_FLASH_OFFSET;
 
 	dma_chunks = image_size / max_dma_size;
diff --git a/common/littlefs/lfs_wrapper.c b/common/littlefs/lfs_wrapper.c
index 9a09985..203279c 100644
--- a/common/littlefs/lfs_wrapper.c
+++ b/common/littlefs/lfs_wrapper.c
@@ -162,14 +162,14 @@
 static int ram_block_read(const struct lfs_config *c __attribute__((unused)), lfs_block_t block,
         lfs_off_t off, void *buffer, lfs_size_t size)
 {
-	memcpy_safe(buffer, DST_UNSAFE, (void *)(CONFIG_BOOTINFO_ADDR + (block * BLOCK_SIZE) + off), size);
+	memcpy_safe(buffer, DST_UNSAFE, (void *)(((uint32_t)CONFIG_BOOTINFO_ADDR) + (block * BLOCK_SIZE) + off), size);
 	return 0;
 }
 
 static int ram_block_prog(const struct lfs_config *c __attribute__((unused)), lfs_block_t block,
         lfs_off_t off, const void *buffer, lfs_size_t size)
 {
-	memcpy_safe((uint8_t *)(CONFIG_BOOTINFO_ADDR + (block * BLOCK_SIZE) + off), DST_UNSAFE, (void *)buffer, size);
+	memcpy_safe((uint8_t *)(((uint32_t)CONFIG_BOOTINFO_ADDR) + (block * BLOCK_SIZE) + off), DST_UNSAFE, (void *)buffer, size);
 	return 0;
 }
 
diff --git a/include/boot_rom.h b/include/boot_rom.h
index 13111fd..e9e75ac 100644
--- a/include/boot_rom.h
+++ b/include/boot_rom.h
@@ -9,9 +9,9 @@
 #include <stdio.h>
 #include <stdint.h>
 #include <stdbool.h>
-#include "hl_boot_desc_rom.h"
+#include <hl_boot_desc_rom.h>
 #include <hl_sec_if.h>
-#include "spi.h"
+#include <spi.h>
 
 #define wfe()	__asm__ __volatile__("sleep 0x11\n" : : )
 
@@ -29,17 +29,33 @@
 
 #define BIT(nr)					(1 << (nr))
 
+#define GPIOS_PER_BANK			16
+
 #define ARRAY_SIZE(x)				(sizeof(x) / sizeof((x)[0]))
 /* extracts "n" number of bits from position "p" in the given number "v" */
 #define EXTRACT_BITS(v, p, n)	(((v) & (((1 << (n)) - 1) << (p))) >> (p))
 
 #define CRC32_SIZE		4
 #define SIZE_KB			1024
-#define SIZE_MB			(SIZE_KB * SIZE_KB)
+#define SZ_16				0x00000010
+#define SZ_64K				0x00010000
+#define SZ_128K				0x00020000
+#define SZ_1MB				(SIZE_KB * SIZE_KB)
+#define SZ_2M				0x00200000
+#define SZ_4MB				0x00400000
+#define SZ_256MB			0x10000000
+#define SZ_1G				0x40000000
 
 #define STATIC_DIE_TYPE_IO				0x5
 #define STATIC_DIE_TYPE_COMP				0xA
 
+enum gpio_func {
+	GPIO_FUNC_0,
+	GPIO_FUNC_1,
+	GPIO_FUNC_2,
+	GPIO_FUNC_3,
+};
+
 enum boot_rom_manifests {
 	BOOT_ROM_PRIMARY = 0,
 	BOOT_ROM_SECONDARY = 1,
diff --git a/include/mem_io.h b/include/mem_io.h
index 81fef8d..84a812f 100644
--- a/include/mem_io.h
+++ b/include/mem_io.h
@@ -10,17 +10,12 @@
 #include <soc.h>
 #include <spi.h>
 
-#define HL_FLASH_BASE				(HL_GEN_MEM_BASE + HL_FLASH_OFFSET)
-#define HL_SBROM_BASE				(HL_FW_MEM_BASE + HL_SBROM_OFFSET)
-#define HL_FWSRAM_BASE				(HL_FW_MEM_BASE + HL_FWSRAM_OFFSET)
+#define ARC_REGION_UPD(REG_O, ARC_REGION)	((REG_O & ~(ARC_REGION_MASK)) | ARC_REGION) /* TODO: REGION Update should be based on 4mg group */
+
 
 /* In ARC address manipulation BIT[22:24] are for die id */
 #define HL_ARC_ADDR_DIE_ID_SHIFT		25
 #define HL_ARC_ADDR_DIE_ID_MASK			(0x7 << HL_ARC_ADDR_DIE_ID_SHIFT)
-
-#define ARC_REG_UPD(OFFSET, ARC_REG_BASE)	((OFFSET & ~(ARC_REG_BASE_MASK)) | ARC_REG_BASE)
-
-#define MASK_32_BIT				0xFFFFFFFF
 
 #define DIE_BASED_ADDRESS(offset, die_id)	(offset & ~(HL_ARC_ADDR_DIE_ID_MASK)) | \
 				(die_id << HL_ARC_ADDR_DIE_ID_SHIFT)
@@ -39,26 +34,26 @@
 
 static inline uint32_t regmap_readl(uint32_t offset)
 {
-	uint32_t reg = base_readl(ARC_REG_UPD(offset, HL_CFG_BASE));
+	uint32_t reg = base_readl(ARC_REGION_UPD(offset, CFG_SPACE_REG_BASE));
 
 	return reg;
 }
 
 static inline void regmap_writel(uint32_t val, uint32_t offset)
 {
-	base_writel(val, (ARC_REG_UPD(offset, HL_CFG_BASE)));
+	base_writel(val, (ARC_REGION_UPD(offset, CFG_SPACE_REG_BASE)));
 }
 
 static inline uint8_t regmap_readb(uint32_t offset)
 {
-	uint32_t reg = *((volatile uint32_t *)(HL_CFG_BASE + offset));
+	uint32_t reg = *((volatile uint32_t *)(CFG_SPACE_REG_BASE + offset));
 
 	return (uint8_t)(reg & 0xFF);
 }
 
 static inline void regmap_writeb(uint8_t val, uint32_t offset)
 {
-	*((volatile uint32_t *)(HL_CFG_BASE + offset)) = (uint32_t)val;
+	*((volatile uint32_t *)(CFG_SPACE_REG_BASE + offset)) = (uint32_t)val;
 }
 
 static inline void spi_dw_writel(struct spi_ctx *ctx, uint32_t val, uint32_t offset)
